polkadot/node/network/approval-distribution/src/lib.rs

```diff
@@ -1394,57 +1614,32 @@ impl State {
 			in_topology || knowledge.sent.contains(message_subject, MessageKind::Assignment)
 		};
 
-		let (v1_peers, v2_peers) = {
-			let peer_data = &self.peer_data;
-			let peers = entry
-				.known_by
-				.iter()
-				.filter_map(|(p, k)| peer_data.get(&p).map(|pd| (p, k, pd.version)))
-				.filter(|(p, k, _)| peer_filter(p, k))
-				.map(|(p, _, v)| (*p, v))
-				.collect::<Vec<_>>();
-
-			// Add the metadata of the assignment to the knowledge of each peer.
-			for (peer, _) in peers.iter() {
-				// we already filtered peers above, so this should always be Some
-				if let Some(peer_knowledge) = entry.known_by.get_mut(peer) {
-					peer_knowledge.sent.insert(message_subject.clone(), message_kind);
-				}
+		let peers = entry
+			.known_by
+			.iter()
+			.filter(|(p, k)| peer_filter(p, k))
+			.filter_map(|(p, _)| self.peer_views.get(p).map(|entry| (*p, entry.version)))
+			.collect::<Vec<_>>();
+
+		// Add the metadata of the assignment to the knowledge of each peer.
+		for peer in peers.iter() {
+			// we already filtered peers above, so this should always be Some
+			if let Some(entry) = entry.known_by.get_mut(&peer.0) {
+				entry.sent.insert(message_subject.clone(), message_kind);
 			}
-
-			if !peers.is_empty() {
-				gum::trace!(
-					target: LOG_TARGET,
-					?block_hash,
-					?candidate_index,
-					local = source.peer_id().is_none(),
-					num_peers = peers.len(),
-					"Sending an approval to peers",
-				);
-			}
-
-			let v1_peers = filter_peers_by_version(&peers, ValidationVersion::V1);
-			let v2_peers = filter_peers_by_version(&peers, ValidationVersion::V2);
```

Comment:
(not related to this PR, just curious)

Reply:
Because even if ApprovalDistributionMessage is the same between V1 and V2, VersionedValidationProtocol is not, so you would get deserialization failures.
```
pub type VersionedValidationProtocol =
	Versioned<v1::ValidationProtocol, v2::ValidationProtocol, vstaging::ValidationProtocol>;
```
	
```
pub enum Versioned<V1, V2, VStaging> {
	/// V1 type.
	V1(V1),
	/// V2 type.
	V2(V2),
	/// VStaging type.
	VStaging(VStaging),
}
```

Reply:
Not sure, I am following. Why wouldn't something like this be possible:

```
pub enum Versioned<V1, V1, VStaging> {
	/// V1 type.
	V1(V1),
	/// V2 type.
	V2(V1),
	/// VStaging type.
	VStaging(VStaging),
}
```